### PR TITLE
1060 - Signatures for off-chain fees

### PR DESCRIPTION
**Pull request by dysfunctor**
- [x] Set a valid title
- [x] Use branch-name `[<ref>-]<feature-description>`
- [x] Brief description of the changes
- [ ] Create matching PRs in dependent repos

Implements/tests:
- Function for client to sign off-chain submission requests
- Signature verification in Typescript and smart contract
- Starter off-chain fee contract

### CODE REVIEW

File: upa/src/sdk/client.ts [hunk 1]
```diff
@@ -139,3 +142,104 @@ export class UpaClient {
     return this.upaInstance.verifier.registerVK(vk.solidity());
   }
 }
+
+export type UnsignedOffChainSubmissionRequest = {
```
Comment (dtebbs) [inline upa/src/sdk/client.ts]: To note, this conflicts a bit with what's happening in https://github.com/NebraZKP/upa/pull/49.

Let's discuss and coordinate.

File: upa/src/sdk/client.ts [hunk 2]
```diff
@@ -139,3 +142,104 @@ export class UpaClient {
...
+// Signed request to be sent out-of-band to an aggregator
+export type SignedOffChainSubmissionRequest = {
+  circuitIdProofAndInputs: application.CircuitIdProofAndInputs[];
+  submissionId: ethers.BytesLike;
+  expirationBlockNumber: bigint;
+  submitterNonce: bigint;
+  fee: bigint;
+  totalFee: bigint;
+  signature: string;
+};
```
Comment (dtebbs) [inline upa/src/sdk/client.ts]: Is this just:
```suggestion
export type SignedOffChainSubmissionRequest = 
  UnsignedOffChainSubmissionRequest & 
  {  signature: string; };
```

File: upa/src/sdk/client.ts [hunk 3]
```diff
@@ -139,3 +142,104 @@ export class UpaClient {
     return this.upaInstance.verifier.registerVK(vk.solidity());
   }
 }
+
+export type UnsignedOffChainSubmissionRequest = {
+  circuitIdProofAndInputs: application.CircuitIdProofAndInputs[];
+  submissionId: ethers.BytesLike;
```
Comment (dtebbs) [inline upa/src/sdk/client.ts]: Let's use an explicit string

File: upa/src/sdk/client.ts (hunk 1 above)
Comment (dtebbs) [reply upa/src/sdk/client.ts]: (We'll probably want a proper class with `from_json` initializers to thoroughly check the serialized data).  Let's combine with the above.  Maybe we can merge branches.

File: upa/src/sdk/client.ts [hunk 4]
```diff
@@ -139,3 +142,104 @@ export class UpaClient {
...
+  expirationBlockNumber: bigint;
+  submitterNonce: bigint;
+  fee: bigint;
+  totalFee: bigint;
+  signature: string;
+};
+
+// Data to be signed by the requester
+export type SignedRequestData = {
+  submissionId: ethers.BytesLike;
```
Comment (dtebbs) [inline upa/src/sdk/client.ts]: string

File: upa/test/upaTests.ts [hunk 5]
```diff
@@ -804,8 +806,9 @@ describe("UPA", async () => {
     });
 
     it("submitMultipleCheckReceipts", async () => {
-      const { upa, s1, s2, s3, s1_tx, s2_tx, s3_tx } =
-        await loadFixture(deployAndSubmit);
+      const { upa, s1, s2, s3, s1_tx, s2_tx, s3_tx } = await loadFixture(
```
Comment (dtebbs) [inline upa/test/upaTests.ts:809]: There are a tonne of formatting changes which inflate the PR and make it harder to review.
Do you know which version of prettier you are using?

File: upa/contracts/UpaOffChainFee.sol [hunk 6]
```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+
+import {EIP712} from "@openzeppelin/contracts/utils/cryptography/EIP712.sol";
+import {ECDSA} from "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+
+/// Contract that off-chain aggregators may use for handling fee deposits,
+/// payments, and claims.
+contract UpaOffChainFee is EIP712 {
```
Comment (dtebbs) [inline upa/contracts/UpaOffChainFee.sol]: "Deposit" contract?

Comment (dysfunctor) [reply upa/contracts/UpaOffChainFee.sol]: Renamed

File: upa/src/sdk/client.ts (hunk 4 above)
Comment (dysfunctor) [reply upa/src/sdk/client.ts]: done

File: upa/src/sdk/client.ts (hunk 1 above)
Comment (dysfunctor) [reply upa/src/sdk/client.ts]: turned into classes

File: upa/src/sdk/client.ts (hunk 3 above)
Comment (dysfunctor) [reply upa/src/sdk/client.ts]: done

File: upa/src/sdk/client.ts (hunk 2 above)
Comment (dysfunctor) [reply upa/src/sdk/client.ts]: Yeah this is better. Though now that it's a class, I'd want `OffChainSubmissionRequest` to extend `UnsignedOffChainSubmissionRequest`, but it is already extending `OffChainSubmission`. Maybe we just store the whole `OffChainSubmissionRequest` in the DB to simplify the types by removing `OffChainSubmission`?

**Pull request merged**